receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/unmarshaler.go

```diff
@@ -0,0 +1,97 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cwmetricstream // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/unmarshaler/cwmetricstream"
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/collector/model/pdata"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/unmarshaler"
+)
+
+const (
+	Encoding        = "cwmetrics"
+	recordDelimiter = "\n"
+)
+
+var (
+	errInvalidRecords = errors.New("record format invalid")
+)
+
+type Unmarshaler struct {
+}
+
+var _ unmarshaler.MetricsUnmarshaler = (*Unmarshaler)(nil)
+
+func NewUnmarshaler() *Unmarshaler {
+	return &Unmarshaler{}
+}
+
+func (u Unmarshaler) Unmarshal(records [][]byte) (pdata.Metrics, error) {
+	builders := make(map[string]*resourceMetricsBuilder)
+	for _, record := range records {
+		// multiple metrics in each record separated by newline character
+		for _, datum := range bytes.Split(record, []byte(recordDelimiter)) {
+			if len(datum) > 0 {
+				var metric cWMetric
+				err := json.Unmarshal(datum, &metric)
+				if err != nil || !u.isValid(metric) {
+					continue
```

Comment:
Should this be logged?  I'd hate to lose visibility into why some metrics were not processed.

Reply:
I'm not really sure how to log this. Would I log the entire `datum` as a string? This should only fail if the record isn't a CloudWatch metric. Like if the Firehose was for a different format. If all of the records in the request aren't CloudWatch metrics, then it'll return a `record format invalid` error. The fields it is checking will always be present in a metric.

Reply:
If `err != nil` it is probably fine to do something like this:

```go
u.logger.Info("unable to unmarshal input", zap.Error(err), zap.Int("datum_index", di), zap.Int("row_index", ri))
```

Not sure how useful the index values would be, but if the user has access to the request it might help them pinpoint which particular point was failing.

One important consideration here is that the user on the sending side, with access to HTTP responses, may not be the same as the operator of the collector instance generating the responses.  Without logging on the collector side there will be no way for a collector operator to know that anything is going wrong or attempt to guide resolution.